dbms/src/Storages/BackgroundProcessingPool.cpp

```diff
@@ -133,7 +133,10 @@ BackgroundProcessingPool::~BackgroundProcessingPool()
 {
     try
     {
-        shutdown = true;
+        {
+            std::unique_lock lock(tasks_mutex);
+            shutdown = true;
```

Comment:
Can you elaborate about why not acquiring a lock on `tasks_mutex` here will cause a deadlock?

Reply:
Added in the pr description.